bokehjs/src/lib/models/annotations/toolbar_panel.ts

```diff
@@ -44,54 +43,71 @@ export class ToolbarPanelView extends AnnotationView {
 
   override connect_signals(): void {
     super.connect_signals()
+
     this.plot_view.mouseenter.connect(() => {
       this.toolbar_view.set_visibility(true)
     })
     this.plot_view.mouseleave.connect(() => {
       this.toolbar_view.set_visibility(false)
     })
+
+    this.plot_view.canvas.ui_event_bus.focus.connect(() => {
+      this.toolbar_view.toggle_auto(true)
+    })
+    this.plot_view.canvas.ui_event_bus.blur.connect(() => {
+      this.toolbar_view.toggle_auto(false)
+    })
+
+    // TODO this needs to be implemented higher up
+    const {visible} = this.model.properties
+    this.on_change(visible, () => {
+      display(this.el, this.model.visible)
+    })
   }
 
   override remove(): void {
     this.toolbar_view.remove()
     super.remove()
   }
 
+  // TODO remove this when CSS positioning is fully implemented
+  override update_position(): void {
+    if (this.model.visible) {
+      const {x, y, width, height} = this.layout.bbox
```

Comment:
Can these values ever be NaN?

Reply:
No in the case of layout bboxes. In general there can be invalid `BBox` instances (one can check for that with `bbox.is_valid`). However, there's logic higher up in base classes, that prevent rendering/painting to go that far if bbox is invalid.